### PR TITLE
multi: Add ability to handle protocol-range messages in Custom Message APIs

### LINKED ISSUE #7093: [feature]: explicit exceptions for CustomMessage API
**Is your feature request related to a problem? Please describe.**

Today we have the custom message API, but it can only be used to send messages above the custom message range (65k+). If we were allowed to specify some explicit messages that could be sent as an exception, then we'd be able to use them for certain classes of itests. 

As an example: test that if we get a warning message we never actually use it to force close a channel. More advanced use cases would be capturing/dropping protocol messages to retransmit selectively in an itest. Other types of network proxies could also be built using this. 

**Describe the solution you'd like**

Add a new `--protocol` flag that lets a user specify which message type (range or explicit value) that the custom message API should be able to send with. 

**Describe alternatives you've considered**

More generic network/message level hooks. The proposal above should be pretty minimal implementation wise. This also unlocks people implementing other spec drafts/proposals outside of lnd.

### PR DESCRIPTION
**Pull request by carlaKC**
## Change Description
This PR adds the handling of protocol-range messages via the existing custom mesage APIs. This functionality is protected behind the `dev` built tag, and must be opted-into on a per-message level because it is an advanced features. 

Fixes #7093. 

#### Implementation Node
This approach uses a simple global of overriden message numbers to set our list of message types that can be sent with the custom messaging API. Since we only write to this value on startup (and have no cause to write to it otherwise), I went with the simplicity of this approach. Threading through to [makeEmptyMessage](https://github.com/lightningnetwork/lnd/blob/master/lnwire/message.go#L177) would otherwise require a much larger refactor. 

## Steps to Test
Test sending a protocol level message from Alice to Bob. 

#### Without Opting-In
* `lncli-bob subscribecustom`
* `lncli-alice sendcustom --type=555` 
  * Alice: send fails due to type number
  * Bob: does not receive anything (nothing sent)

#### Sender Opt-In
* `lncli-bob subscribecustom`
* Alice: Build lnd with `dev` tag and run with `protocol.custom-message=513` config option
* `lncli-alice sendcustom --type=555`
  * Alice: send should not succeed
  * Bob: does not receive anything (not opted into receiving this message on custom API)

#### Sender and Recipient Opt-in
* Bob: Build lnd with `dev` tag and run with `protocol.custom-message=513` config option
* `lncli-bob subscribecustom`
 * `lncli-alice sendcustom --type=555`
  * Alice: send should not succeed
  * Bob: does not receive anything (not opted into receiving this message on custom API)

### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

File: lnwire/custom.go
Comment (morehouse) [inline lnwire/custom.go:12]: Nit: looks like `CustomTypeStart` should be const.

File: lnwire/custom.go
Comment (morehouse) [inline lnwire/custom.go:33]: We're passing in `uint64` and casting to `MessageType` (`uint16`).  This could result in unexpected truncation (e.g., 65536 becomes 0).

We should probably either pass in `MessageType` or return an error if the type is greater than 65535.

File: lnwire/custom.go
Comment (morehouse) [inline lnwire/custom.go:23]: Perhaps a `map[MessageType]struct{}` would be a more natural data structure to contain the set.

Then we wouldn't need `IsCustomOverride` at all.

File: lntest/itest/lnd_custom_message.go
Comment (Crypt-iQ) [inline lntest/itest/lnd_custom_message.go:35]: type = 554?

File: lntest/itest/lnd_custom_message.go
Comment (Crypt-iQ) [inline lntest/itest/lnd_custom_message.go:166]: message type 1 was delivered?

File: lntest/itest/lnd_custom_message.go
Comment (Crypt-iQ) [inline lntest/itest/lnd_custom_message.go:179]: custom message 1?

**Comment by Crypt-iQ** [comment]
It seems like there's a flake in the itest https://github.com/lightningnetwork/lnd/actions/runs/3464262868/jobs/5785582571

File: lnwire/custom.go
Comment (Roasbeef) [inline lnwire/custom.go:74]: Seems like an alternative path here would be to:
  * add a functional option arg to `NewCustom`
  * let that pass in the set of custom override types
  * do this check based on that 

I think that would let us get rid of the current global that isn't safe for concurrent access?

File: lnwire/custom.go
Comment (Roasbeef) [inline lnwire/custom.go:44]: Accesses a global map w/o any synchronization.

File: lnwire/custom.go
Comment (ellemouton) [inline lnwire/custom.go:77]: So this causes any custom types not defined in `lnwire/messages.go` to be displayed here as `<unknown>`. But since we are definitely going to always have types not defined yet, should we maybe print out the type numbers here rather?

**Review by ellemouton — COMMENTED** [review]
tACK! Slick PR 🔥 

Main comment from me is about the the error message returned when a message is not in the list of defined allowed overrides.

File: lnwire/custom.go
Comment (carlaKC) [reply lnwire/custom.go:12]: Pre-existing (this is a var in master) - will pull it out into a const in refactor commit.

File: lnwire/custom.go
Comment (carlaKC) [reply lnwire/custom.go:74]: > pass around the list of allowed overrides everywhere then? That might be a bit tricky (and add lots of refactoring)

Yeah, this is the problem with trying to move away from a global - lots of lift in various parts of the code base.

File: lnwire/custom.go
Comment (carlaKC) [reply lnwire/custom.go]: Added in c89dce6.

File: lnwire/custom.go
Comment (carlaKC) [reply lnwire/custom.go]: iiuc, you mean use `%d` rather than `%v` in the error? 
If yes, updated! If no, not sure what you mean here.

File: lnwire/custom.go
Comment (carlaKC) [reply lnwire/custom.go]: Went with a simple RW mutex + allowing the ability to call `SetCustomOverrides` rather than erroring out on multiple calls. Do you think we need to be that strict?

**Comment by carlaKC** [comment]
> It seems like there's a flake in the itest https://github.com/lightningnetwork/lnd/actions/runs/3464262868/jobs/5785582571

Lost the logs here because GH expires them after 5 days :( I suspect this will be one of those frustrating timeout things, custom messages are queued with low priority so can sometimes take pretty long to arrive - but will dig into fresh logs as the flakes appear!

**Comment by carlaKC** [comment]
Added synchorinization to custom overrides global + addressed majoy comments - ready for another look! 

Still digging into the cause of the itest flake (can't reproduce locally with flake hunter, only seems to be appearing on "old" itests?).

File: lnwire/custom.go
Comment (Crypt-iQ) [inline lnwire/custom.go:86]: I think this usage of `customTypeOverride` needs the read lock

File: lnwire/custom.go
Comment (Crypt-iQ) [inline lnwire/custom.go:62]: I don't think this can be hit since `SetCustomOverrides` is always called and will initialize `customTypeOverride`

File: lntest/itest/lnd_custom_message.go
Comment (Crypt-iQ) [inline lntest/itest/lnd_custom_message.go:164]: nit: type 1?

**Review by Crypt-iQ — COMMENTED** [review]
I think this is almost there, I haven't been able to reproduce the flake locally either

File: lnwire/custom.go
Comment (carlaKC) [reply lnwire/custom.go:86]: Thought I'd removed it, I don't think it's worth printing the whole map every time so simpler without it.

File: lnwire/custom.go
Comment (carlaKC) [reply lnwire/custom.go:62]: I left this in to safeguard against a panic if we call `IsCustomOverride` before `SetCustomOverrides`, but I guess that's desired behavior because we should never call the check before the we set (which happens in config parsing, so really should never happen)?

File: lnwire/custom.go
Comment (ellemouton) [reply lnwire/custom.go]: will there ever be a case where we want to call `SetCustomOverrides` more than once throughout the lifetime of lnd? perhaps right? like if we wanted to add a new allowed custom message type via lncli and not need to restart lnd. In that case, this method will need to be changed so that it appends to the list of overrides & does not keep overwriting it. 
Non blocking though. Just a thought.

File: lnwire/custom.go
Comment (ellemouton) [reply lnwire/custom.go]: yeah that is what I meant :)

**Review by ellemouton — APPROVED** [review]
LGTM! 🔥 🔥   just needs a one more push or rebase so that the linter fix can take hold.

Left one comment about perhaps adding to the list of overrides instead of overwriting the list. Non-blocking imo

File: lnwire/custom.go
Comment (ellemouton) [reply lnwire/custom.go]: sounds gooood

**Comment by carlaKC** [comment]
Test still flaking, will dig into that properly before merge.

**Review by Crypt-iQ — APPROVED** [review]
The flake still exists - see the bitcoind-etcd itest. I can't figure out why it happens. LGTM once that's fixed

**Comment by carlaKC** [comment]
Itest flake has been hunted! The problem was that the message from Bob was being sent before Alice was properly subscribed with `SubscribeCustomMessage`, so when the message arrives there's no subscription client to see it. 

The solution is just to wait for alice to be fully booted up after restart, then subscribe, then send the message.

File: lnwire/custom.go
Comment (Roasbeef) [reply lnwire/custom.go:78]: One way to do this would be to create a new struct something like `CustomMsgWriter/Reader`. Then that can accept the set of new custom message types that should be permitted. Relevant call sites that need to be able to handle them would then instantiate this, and use it instead of the direct `lnwire.Read/WriteMessage` calls. `lnwire.WriteMessage` and the like can then gain the ability to forcibly read/write a custom message out of te range via an optional arg. With this path, there's no global needed and the call sites that don't necessarily need to handle this new custom msg range don't change (functional arg so no ned for them to pass the extra message).

File: lnrpc/lightning.proto
Comment (morehouse) [inline lnrpc/lightning.proto:571]: This description is a little ambiguous for someone unfamiliar with the spec.  Does ">= 32768" refer to types inside or outside the custom range?

```suggestion
    To include messages with type below the custom range (>= 32768) lnd
```

Comment (carlaKC) [reply lnrpc/lightning.proto:571]: I don't think that somebody unfamiliar with the spec _should_ be sending messages that are outside of the cutom range :') but will update the wording if I push again 👍

**Pull request merged**
